### ISSUE TITLE
Controlling optional argument in a resource

### ISSUE THREAD

**Issue by omerh** [hashicorp/terraform#17968]
I am looking for a way to control AWS autoscaling group resource "desired_capacity"
I've built a module that I want to have optionally desired_capacity in it.

something like: (which of course this doesnt work)
```
resource "aws_autoscaling_group" "this" {
  name                      = "${var.name}"
  max_size                  = "${var.group_max_size}"
  min_size                  = "${var.group_min_size}"

  if { set_cap = true ? desired_capacity = 1 : null }

  health_check_grace_period = "${var.health_grace_period}"
  health_check_type         = "${var.health_check_type}"
  launch_configuration      = "${var.launch_configuration_id}"
  vpc_zone_identifier = "${var.vpc_subnets}"
}

```

So only if in the module I have added `set_cap = true` than terraform will keep the state of 1 else it will ignore it.

What am I missing?

**Comment by apparentlymart** [comment]
Hi @omerh! Sorry that this use-case is tricky to meet right now.

The configuration language doesn't currently have this idea of `null` as you illuatrated, and so it's not possible to represent conditionally assigning an attribute like this. For such situations, currently this is often worked around by defining two separate resource blocks and using conditional `count` to ensure that only one is enabled at a time:

```hcl
resource "aws_autoscaling_group" "set_cap" {
  count = "${var.set_cap ? 1 : 0}"

  desired_capacity = 1
}

resource "aws_autoscaling_group" "unset_cap" {
  count = "${var.set_cap ? 0 : 1}"

  # no desired_capacity here
}
```

Obviously this is non-ideal for many reasons, and in particular in this case because changing the value of `set_cap` would cause the autoscaling group to be entirely replaced, rather than updated in-place.

We are currently integrating an improved version of the configuration language interpreter that _does_ have the concept of `null` and will support syntax like the following:

```hcl
resource "aws_autoscaling_group" "set_cap" {
  desired_capacity = var.set_cap ? 1 : null
}
```

This is planned to be included in the next major release. This is one of the ideas being discussed in #14037, but since that is covering a few different ideas and is thus harder to find in search let's treat this issue as representing this particular feature and we'll post future updates about it here as we get closer to release.

**Comment by omerh** [comment]
@apparentlymart thanks a lot for the informative answer.
Ill be tracking this issue, as I see it as super important for my use case of repeatable module usage

**Comment by apparentlymart** [comment]
Hi @omerh!

The change I mentioned before is now merged into master, ready to be included in the forthcoming v0.12.0 release.

I verified this in the v0.12.0-alpha1 prerelease build using the following simplified configuration:

```hcl
provider "aws" {
  region = "us-west-2"
}

variable "desired_capacity" {
  type    = number
  default = null
}

resource "aws_autoscaling_group" "this" {
  name = "example"

  max_size         = 0
  min_size         = 10
  desired_capacity = var.desired_capacity

  launch_configuration = "placeholder"
  vpc_zone_identifier  = ["placeholder"]
}
```

In this case I did something slightly different than what I mentioned before: I just made the variable default to `null` so that `desired_capacity` will be set only if the user explicitly sets the variable.

```
$ terraform plan

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # aws_autoscaling_group.this will be created
  + resource "aws_autoscaling_group" "this" {
      + arn                       = (known after apply)
      + default_cooldown          = (known after apply)
      + desired_capacity          = (known after apply)
      + force_delete              = false
      + health_check_grace_period = 300
      + health_check_type         = (known after apply)
      + id                        = (known after apply)
      + launch_configuration      = "placeholder"
      + load_balancers            = (known after apply)
      + max_size                  = 0
      + metrics_granularity       = "1Minute"
      + min_size                  = 10
      + name                      = "example"
      + protect_from_scale_in     = false
      + service_linked_role_arn   = (known after apply)
      + target_group_arns         = (known after apply)
      + vpc_zone_identifier       = [
          + "placeholder",
        ]
      + wait_for_capacity_timeout = "10m"
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

Because `desired_capacity` gets populated by the provider itself when not explicitly set, it shows up by default as "known after apply" (which is the new terminology for "computed") in this case.

But if I explicity set the variable, it is populated with my constant value:

```
$ terraform plan -var="desired_capacity=3"

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # aws_autoscaling_group.this will be created
  + resource "aws_autoscaling_group" "this" {
      + arn                       = (known after apply)
      + default_cooldown          = (known after apply)
      + desired_capacity          = 3
      + force_delete              = false
      + health_check_grace_period = 300
      + health_check_type         = (known after apply)
      + id                        = (known after apply)
      + launch_configuration      = "placeholder"
      + load_balancers            = (known after apply)
      + max_size                  = 0
      + metrics_granularity       = "1Minute"
      + min_size                  = 10
      + name                      = "example"
      + protect_from_scale_in     = false
      + service_linked_role_arn   = (known after apply)
      + target_group_arns         = (known after apply)
      + vpc_zone_identifier       = [
          + "placeholder",
        ]
      + wait_for_capacity_timeout = "10m"
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

Since this change is already merged, I'm going to close this out now. It'll be included in the forthcoming v0.12.0 final release. Thanks for reporting this, and thanks for your patience while we got this done.

**Comment by carolinux** [comment]
I will be super happy when this is publicly available
